### ISSUE TITLE
java.net.MalformedURLException: Unknown protocol: ws

### ISSUE THREAD

**Issue by Zo2m4bie** [socketio/socket.io-client-java#233]
Hi. i try to connect to socket server which start with "ws://..." and when I set this url to 

```
 IO.socket(pathSocket, options);
```

I receive error

```
 java.lang.RuntimeException: java.net.MalformedURLException: Unknown protocol: ws
 W/System.err﹕ at io.socket.client.Url.parse(Url.java:52)
 W/System.err﹕ at io.socket.client.IO.socket(IO.java:60)
 W/System.err﹕ at io.socket.client.IO.socket(IO.java:41)
```

**Comment by nkzawa** [comment]
Ah, I didn't know `URL` doesn't accept `ws`. It's a bug, please use `http://` url for now.

**Comment by Zo2m4bie** [comment]
Also the same problem with wss://

**Comment by Zo2m4bie** [comment]
Can I wait this fix soon?

**Comment by Zo2m4bie** [comment]
I wanted t ofix it by myself but there're not all class in project

**Comment by nkzawa** [comment]
Just use `http://` or `https://`. They are actually the same with `ws://` and `wss://` .

**Comment by SteveMobile** [comment]
we use the websocket can connect wss:// , but use this library ,can't connect the "https://" , no any reponse,can you help ?

**Comment by PuKoren** [comment]
+1, this is causing heavy issues on socket servers: losing a lot of CPU on back and forth requests upgrading the connexion. This is a major issue when having 10k connection/second

**Comment by kruyvanna** [comment]
+1

**Comment by sindrenm** [comment]
This is caused by java.net.URL not knowing how to open a connection towards a `ws://` stream. It could be solved by using java.net.URI instead, which is what the [Java-WebSocket library uses][uri].

[uri]: https://github.com/TooTallNate/Java-WebSocket/blob/58d17786958ec2b629b75de1eed00915b3e4f7c4/src/main/java/org/java_websocket/client/WebSocketClient.java#L37

**Comment by sr71k** [comment]
did  it  fix?

**Comment by Agraphie** [comment]
are there any news on this?

**Comment by b95505017** [comment]
Maybe we could use this to resolve the issue
https://medium.com/square-corner-blog/okhttps-new-url-class-515460eea661#.v55accs1c

**Comment by jvilya** [comment]
We are using Socket.io 1.5 and this library on Android client. For connection string we are using http(s), but it switches to ws automatically with these setup:

```java
    IO.Options opts = new IO.Options();
    opts.transports = new String[]{ WebSocket.NAME};
    return IO.socket("https://socket.io.server", opts);
```

**Comment by isanwenyu** [comment]
@jvilya hi,I can't find the 1.4.5 release version in this repo.I  have tried your code above,but it doesn't work.
Could you tell me details,please.

**Comment by jvilya** [comment]
@isanwenyu yes. Try to debug it with any proxy-debugging tool and see what happens
In my case I saw  connection to `https` and then switch to `ws`

**Comment by nuttyboysoftware** [comment]
2 Years later and still no fix? is this product no longer being supported?

**Comment by maheshputtegowda1990** [comment]
hello guys this ws: issue is fixed?

**Comment by maheshputtegowda1990** [comment]
any solution to implement

**Comment by GirishBhutiya** [comment]
Still not working and it is not connect with http:// .

**Comment by Kiplol** [comment]
I'm having this issue as well, blocking development of a feature for our Android app.

**Comment by PhonegapProjects** [comment]
not working at all!!!!!!!!!

**Comment by jakubzieba** [comment]
after a year ... it still does not work

**Comment by songkadi** [comment]
no hope for this topic

**Comment by tiholic** [comment]
https://github.com/infitio/flutter_socket_io/issues/34

**Comment by Arnique** [comment]
The socket server I'm connecting to only works with wss...any solutions for 2019?

**Comment by xiaoshengmm** [comment]
java.lang.RuntimeException: java.net.MalformedURLException: Unknown protocol: ws   who know？how to fix it

**Comment by rogerthat94** [comment]
> Maybe we could use this to resolve the issue
> https://medium.com/square-corner-blog/okhttps-new-url-class-515460eea661#.v55accs1c

This doesn't fix the ws: issue. OkHttp just replaces ws: with http: before parsing the URL.

**Comment by felipecsl** [comment]
from the `HttpUrl` javadocs:
```
 * Sometimes referred to as *protocol*, A URL's scheme describes what mechanism should be used to
 * retrieve the resource. Although URLs have many schemes (`mailto`, `file`, `ftp`), this class only
 * supports `http` and `https`. Use [java.net.URI][URI] for URLs with arbitrary schemes.
```

**Comment by H4kt** [comment]
Idk if someone is still looking for an easy hot-fix, but I'll just leave it here.
`
    public static void addURLProtocols() {

        try {

            Field field = URL.class.getDeclaredField("handlers");
            field.setAccessible(true);

            URLStreamHandler handler = new URLStreamHandler() {
                @Override
                protected URLConnection openConnection(URL url) throws IOException {
                    return new URLConnection(url) {
                        @Override
                        public void connect() throws IOException {}
                    };
                }
            };

            Hashtable<String, URLStreamHandler> handlers = (Hashtable<String, URLStreamHandler>) field.get(null);
            handlers.put("wss", handler);
            handlers.put("ws", handler);

        } catch(Exception ex) {
            ex.printStackTrace();
        }

    }
`

**Comment by vasilevungureanu** [comment]
Hi,

I was able to solve this problem in this way:

`
 val options = IO.Options()
            options.forceNew = true
            options.path = "/wss"

 socket = IO.socket("https://", options)
           
`

**Comment by Mohamed-Shalabi** [comment]
Nothing here is working for me..
I have used org.java_websocket and it worked well, but I need to know what is the problem with socket io!

**Comment by H4kt** [comment]
@Mohamed-Shalabi 
Have you tried the code below?

> Idk if someone is still looking for an easy hot-fix, but I'll just leave it here.
> 
> ```
>     public static void addURLProtocols() {
> 
>         try {
> 
>             Field field = URL.class.getDeclaredField("handlers");
>             field.setAccessible(true);
> 
>             URLStreamHandler handler = new URLStreamHandler() {
>                 @Override
>                 protected URLConnection openConnection(URL url) throws IOException {
>                     return new URLConnection(url) {
>                         @Override
>                         public void connect() throws IOException {}
>                     };
>                 }
>             };
> 
>             Hashtable<String, URLStreamHandler> handlers = (Hashtable<String, URLStreamHandler>) field.get(null);
>             handlers.put("wss", handler);
>             handlers.put("ws", handler);
> 
>         } catch(Exception ex) {
>             ex.printStackTrace();
>         }
> 
>     }
> ```

**Comment by darrachequesne** [comment]
For future readers: this should be fixed by https://github.com/socketio/socket.io-client-java/commit/67fd5f34a31c63f7884f82ab39386ad343527590, included in version `2.0.1`.

**Comment by TheFoLLoNeRoU** [comment]
> For future readers: this should be fixed by [67fd5f3](https://github.com/socketio/socket.io-client-java/commit/67fd5f34a31c63f7884f82ab39386ad343527590), included in version `2.0.1`.

Thank you very mach!! I was using version very old and i was founding very problems with exception about "ws" and problems of connected to server with protocol "http"
